api/xds_protocol.rst

```diff
@@ -469,3 +469,155 @@ occurred via a :ref:`resource update <xds_protocol_resource_update>`.
+resources field will be added to the route configuration maintained by Envoy.
+If Envoy's route configuration already contains a given virtual host, it will
+be overwritten by data received in the
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+but only if the updated virtual host is different from its current state.
+The :ref:`resource.aliases <envoy_api_field_resource.aliases>` field contains
+all host/authority header values used to create the on-demand request. The
+configuration server needs to keep track of virtual hosts known to Envoy.
+
+If a virtual host is requested for which the management server does not know
+about, then the management server should respond with a 
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>` in which
+the :ref:`resources <envoy_api_field_DeltaDiscoveryResponse.resources>` entry
+for that virtual host has the :ref:`name <envoy_api_field_resource.name>` and
+:ref:`aliases <envoy_api_field_resource.aliases>` set to the requested host
+entry and the resource unpopulated. This will allow Envoy to match the
+requested resource to the response.
+
+Envoy should respond to the 
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+with an Ack/Nack 
+:ref:`DeltaDiscoveryRequest's <envoy_api_msg_DeltaDiscoveryRequest>`.
+
+.. figure:: diagrams/delta_rds_request_additional_resources.svg
+   :alt: Delta RDS request additional resources
+
+Request Path Resolution
++++++++++++++++++++++++
+If a route for the contents of the host/authority header cannot be resolved:
+- A :ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>` as 
+described above is queued for transmission.
+- A callback resuming the decoder filter chain of the current active stream is
+created. Together with the callback the contents of all host/authority headers
+used in the request is stored. 
+- The decoder filter chain of the current active stream is paused.
+If there's already a route available, the control is passed to the next filter
+in the filter chain.
+
+When a :ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+arrives:
+- Route configuration is updated.
+- All callbacks whose list of host/authority header values exactly matches the
+list in the :ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+are triggered.
+- The decoder filter chain is resumed. If a route for the host/authority header
+can be found, the active stream is recreated (to pick up the updated route
+configuration). If there's still no route, the control is passed to the next
+filter in the filter chain.
+  
+Unsubscribing from Virtual Hosts
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The xDS management server will also support the ability of Envoy to tell it
+when a resource hasn't been used and is safe to stop monitoring. The resources
+that can be removed include the base resources that the xDS management server
```

Comment:
From an xds protocol perspective, this seems very strange to me.  The virtual hosts initially sent to the client in the RDS response are not individual resources; they're just data in a repeated field within the route configuration.  The client is not actually subscribed to these resources via VHDS, so sending an unsubscription seems inappropriate.

Similarly, if a virtual host is included in the initial route configuration, I would not expect the client to see a VHDS response containing an update for that virtual host, since that is not a resource that the client subscribed to via VHDS.  (I don't see anywhere that this doc explicitly states that this would happen, but if you are thinking that these resources can be unsubscribed to, maybe you are also thinking that they can receive spontaneous updates via VHDS.)

I think that if a virtual host is included in the RDS response, the only way to update that virtual host should be to send an RDS update.

Reply:
The idea here is that an Envoy proxy joins a control plane and receives an initial set of virtual hosts that may be useful to it. For instance, if all of the services are in the same Kubernetes namespace, there might be a higher likelihood that they would talk to each other than services spread across multiple Kubernetes namespaces. These resources may change on the control plane side, and Envoy should be able to update with those changes. If a resource is found not to be used, however, then there's no reason for Envoy to track it and continue receiving updates for it and an unsubscribe can be sent. 

The unsubscribe functionality comes from the delta xDS protocol and is currently not implemented, but I think the rest of this is already implemented in Envoy.

Reply:
VHDS is an optional API on top of RDS, if VHDS is enabled, then communication of virtual hosts for that route configuration changes from being over RDS to being over VHDS. Not sure if it would make sense for virtual hosts to be sent over two different APIs at the same time.

Reply:
@dmitri-d Can comment from the Envoy implementation side.

Reply:
Another comment is that RDS updates reset the virtual host subscriptions meaning that all virtual hosts will have to be reloaded from the management server. Whereas updating a single virtual host over VHDS will only update that virtual host. 

If we have the original set of virtual hosts only update over RDS changes and not over VHDS, then this means that some virtual host updates will clear all of the virtual hosts and require a full reload while some just update relevant content.

Reply:
Okay, so it sounds like you are actually expecting the client to magically have an implicit VHDS subscription for any virtual host present in the RDS response, even though the client never sent a VHDS subscription message for those resources.  Because it is assumed to have subscribed to those resources (even though it never sent a subscription request for them), it can receive updates via VHDS when those resources change, and it can unsubscribe to those resources.

What I'm saying is that, from an xds protocol perspective, I don't think it makes sense to have the VHDS subscriptions for these resources happen magically and implicitly based on the RDS response, since this is basically a layering violation.  From a protocol perspective, VHDS and RDS are supposed to be independent streams, and the state of the VHDS stream should not be affected by the state of the RDS stream, because that would bake that relationship into the protocol -- all xds clients would need to support this relationship in order to speak this protocol.

I can think of a few possible alternatives here:

- We could say that the client should explicitly send VHDS subscription messages for any virtual host that is present in the RDS response.  This is slightly sub-optimal in that it winds up sending each virtual host to the client twice, but it does avoid the layering violation issue.

- We could have the RDS response include a list of virtual hosts for the client to automatically subscribe to via VHDS, and the client would send a VHDS subscription request as soon as it got the RDS response.  This would cause one more round-trip before the data gets to the client on startup, but it would avoid sending the data twice.

- If we are not willing to suffer either sending the data to the client twice or waiting for the extra round-trip at startup, then we could consider some way for the VHDS request to tell the server that it should subscribe to changes but not to re-send the same version of the virtual host data that the client already received in the RDS response.  For example, we could provide a version with each virtual host in the RDS response, and then pass that version in the VHDS subscription request.  (There are a lot of details that would need to be worked out here, but we can get into them if we want to pursue this.)

@htuch may want to weigh in on this as well.

Reply:
> Another comment is that RDS updates reset the virtual host subscriptions meaning that all virtual hosts will have to be reloaded from the management server. Whereas updating a single virtual host over VHDS will only update that virtual host.
> 
> If we have the original set of virtual hosts only update over RDS changes and not over VHDS, then this means that some virtual host updates will clear all of the virtual hosts and require a full reload while some just update relevant content.

Why is it necessary to clear out the full list of virtual hosts whenever the route configuration changes?  This seems like yet another layering violation -- the VHDS stream is already subscribed to a certain set of resources, and VHDS should not automatically resend the same resources it has already sent just because the route configuration has been updated.  This doc already says that the server should not resend resources that have not changed (and the language around this is being clarified in #9029).

I think a better way of modeling this would be to say that any virtual hosts that are inlined in the route configuration will be updated via RDS and any resources that are fetched via VHDS will be updated via VHDS.  If you need to be able to update virtual hosts without updating the whole route configuration, then you should send them via VHDS instead of RDS.

Reply:
> Why is it necessary to clear out the full list of virtual hosts whenever the route configuration changes

VHDS subscription is logically a child of an RDS subscription, and only makes sense in that context. When parent context changes, it may have implications on VHDS, for example:
 - VHDS may go away
 - VHDS config server may change
 - virtual hosts discovered via VHDS got listed in the RouteConfiguration proper
etc.

Too many edge cases at this point to keep track of, hence the recreation of the VHDS subscription.

Reply:
I think there are reasonable ways of handling all of these cases.

If the RDS update removes the vhds config, then the client can close the VHDS stream and delete any virtual hosts that were obtained via VHDS.

If the VHDS server changes, then we can model it as VHDS going away (as per previous paragraph) and then being re-added (i.e., we start the VHDS stream to the new server and then send subscription messages on that stream as needed).

If we take the approach I suggested above where the client will automatically send a VHDS subscription for each virtual host in the RDS response, then that would also address the case where a virtual host that was previously fetched via VHDS is then added to the RDS response.

Reply:
I think my take on this is that it's valid to have a client do what @dmitri-d describes (simple/safe but less efficient), but it shouldn't be the only way to implement VHDS, i.e. we shouldn't be prescriptive. It seems totally fine to also have clients do what @markdroth suggests, the xDS protocol spec should leave room to enable both and have management servers robust to both kinds of clients.

Reply:
>  I think that if a virtual host is included in the RDS response, the only way to update that virtual host should be to send an RDS update.

It makes sense as at the protocol design level, not sure it makes sense to enforce this though.

Reply:
> I think there are reasonable ways of handling all of these cases.

I'm not disagreeing, but I don't want to complicate the first implementation of VHDS with these, it's pretty involved as is.

Reply:
Yeah, I could see it being dangerous to enforce. For instance, Istio by default, filters to services in the same Kubernetes namespace for the base case. If that contains serverless containers, then that means that every time a serverless container goes down or up,then that gets reset. 

It's also proposed above that a good implementation might be to send all subscribed virtual hosts in the RDS connection when RDS connects. If that happens with subscribed virtual hosts that belong to serverless containers, then that also means that every time the virtual host changes, the RDS gets reset and Envoy will experience VHDS amnesia.

Reply:
> I think my take on this is that it's valid to have a client do what @dmitri-d describes (simple/safe but less efficient), but it shouldn't be the only way to implement VHDS, i.e. we shouldn't be prescriptive. It seems totally fine to also have clients do what @markdroth suggests, the xDS protocol spec should leave room to enable both and have management servers robust to both kinds of clients.

I think that what @dmitri-d describes is not actually simpler -- it actually requires more state to be kept both on the management server and on the clients.  Once Envoy depends on the management server to proactively send VHDS resources that were not explicitly requested, then that means that all management servers need to do that.  This seems like a fairly large burden when we've been trying to prioritize keeping things simple for management server implementations.

If we are actually going to have a generic transport protocol, then that transport protocol needs to have clearly defined semantics.  This proposal violates the semantics that we've agreed upon thus far.  I think that if we allow this, then we are basically saying that there is no hope of cleanly separating the transport protocol from the data model, because the data model will always impact how the transport protocol works.

Reply:
@dmitri-d How about the following:

- We maintain two lists of virtual hosts, one that comes in the RDS response and another that comes from a VHDS stream.
- Whenever a given virtual host definition exists in both VHDS and RDS, the one obtained from VHDS is used, because it is assumed to be more recent.
- The client automatically does a VHDS subscription for any virtual host in the RDS response.
- The virtual hosts obtained via VHDS are not deleted when an RDS update is received, as long as the `vhds` field in the RDS update has not changed.  If the `vhds` field does change, then VHDS stream will be closed and all of the associated virtual hosts will be deleted.  If the `vhds` field has a new value that points to a different server, we create a new VHDS stream to that server and subscribe to virtual hosts the same way that we would at client startup.

That way, we are not breaking the layering by implicitly affecting the state of the transport protocol, and we avoid the complexity of requiring management servers to keep joint state for RDS and VHDS.  The only down-side is that the virtual hosts in the RDS response will be transferred twice when the client starts up, but that seems manageable, and we can optimize later if it becomes a problem.

WDYT?

Reply:
This solves my concern for constantly changing, or limited lifetime resouces in the base config. However, does the third bullet make the first two unnecessary? I.e. the results of the initial VHDS DeltaDiscoveryRequest, if VHDS is enabled,  would replace the initial resources from RDS, meaning that the VHDS list would always replace the RDS list when it's enabled? (or remove them if the resource has gone away)

Reply:
This thread is now quite long and complicated. Looking at Mark's last suggestion, I think the crux of the complexity here is the fact that the same virtual host can be in the RDS and provided by VHDS. Can we just make these inherently disjoint sets?

I would ideally like to be able to describe VHDS operation completely independent of RDS. We shouldn't need to be prescriptive on what client implementation looks like there.

Reply:
@brian-avery Yeah, if we say that the VHDS data just replaces the RDS data, then I agree that having two sets doesn't buy us anything.

That having been said, I think there's still a problem here, because for any virtual host that is present in the RDS response, the management server will still need to send an RDS update when that virtual host changes.  The only way it could *not* do that is if it had special knowledge that VHDS is in use, and I don't think we should require management servers to modify their RDS behavior based on whether VHDS is configured.  The two should really be independent.

I like @htuch's suggestion of requiring that the two sets of virtual hosts be disjoint.  That would just mean that you would have to choose between sending a virtual host in RDS (which would require sending an RDS update whenever that virtual host changes) or sending it in VHDS (which would require an extra round-trip to get the data at startup).  And if we make the change that @dmitri-d suggests above to avoid clearing the VHDS state on an RDS update that does not change the `vhds` field, then the RDS updates should be minimally disruptive.

Reply:
I agree. Separating the two sounds good to me and the change that Dmitri suggested should alleviate the constant clearing.

Reply:
+1 to avoiding clearing VHDS on RDS update where `vhds` is unchanged.